rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -1197,7 +1276,119 @@ impl SystemState {
+        match &mut self.status {
+            // Return the stop context, nothing to do here.
+            CanisterStatus::Stopped => Some(stop_context),
+
+            CanisterStatus::Stopping {
+                ref mut stop_contexts,
+                ..
+            } => {
+                // Add the message so we can respond to it once the canister has fully stopped.
+                stop_contexts.push(stop_context);
+                None
+            }
+
+            CanisterStatus::Running {
+                call_context_manager,
+            } => {
+                // Transition the canister into the stopping state.
+                self.status = CanisterStatus::Stopping {
+                    call_context_manager: std::mem::take(call_context_manager),
+                    // Track the stop message to respond to it once the canister is fully stopped.
+                    stop_contexts: vec![stop_context],
+                };
+                None
+            }
+        }
+    }
+
+    /// Tries to transition a `Stopping` canister into the `Stopped` state. No-op if
+    /// the canister is `Running` or already `Stopped`.
+    ///
+    /// Returns a tuple of:
+    ///  * a boolean indicating whether the canister has stopped,
+    ///  * all stop contexts if the canister has stopped; or the expired stop
+    ///    contexts only if the canister is still stopping.
+    #[must_use]
+    pub fn try_stop_canister(
+        &mut self,
+        is_expired: impl Fn(&StopCanisterContext) -> bool,
+    ) -> (bool, Vec<StopCanisterContext>) {
+        match self.status {
+            // Canister is not stopping so we can skip it.
+            CanisterStatus::Running { .. } | CanisterStatus::Stopped => (false, Vec::new()),
+
+            // Canister is ready to stop.
+            CanisterStatus::Stopping {
+                ref call_context_manager,
+                ref mut stop_contexts,
+            } if call_context_manager.callbacks().is_empty()
+                && call_context_manager.call_contexts().is_empty() =>
+            {
+                let stop_contexts = std::mem::take(stop_contexts);
+
+                // Transition the canister to "stopped".
+                self.status = CanisterStatus::Stopped;
+
+                // Reply to all pending stop_canister requests.
+                (true, stop_contexts)
+            }
+
+            // Canister is stopping, but not yet ready to stop.
+            CanisterStatus::Stopping {
+                ref mut stop_contexts,
+                ..
+            } => {
+                // Return any stop contexts that have timed out.
+                let mut expired_stop_contexts = Vec::new();
+                stop_contexts.retain(|stop_context| {
+                    if is_expired(stop_context) {
+                        expired_stop_contexts.push(stop_context.clone());
+                        false
+                    } else {
+                        true
+                    }
+                });
```

Comment:
Can we maybe in order to avoid cloning, do something like
`let expired_stop_contexts = stop_contexts.extract_if(|stop_context|  is_expired(stop_context));`?

Reply:
That looks like it would be less efficient, on two accounts. Yes, cloning is potentially more expensive than moving, but:
 1. This code clones every stop context at most once.
 2. I would expect the vast majority (if not all) of stop contexts to stay put most of the time. Whereas taking apart the `Vec` and putting it back together in a newly allocated memory location every round is a lot more work.

Worst-case scenario with this code is that the first stop context expires on every call; it needs to be cloned; and the remaining stop contexts all shifted forward by one. But apart from the one clone, this is the average case with the `partition()` approach (plus needing to allocate at least one new `Vec`). Whereas the average case here is that nothing is mutated.

Reply:
I agree with the last passage, but I think the worst case is when all stop contexts expire. But if that is not expected, feel free to ignore it.

Reply:
As said, every stop context will expire at most once. Sure, they may expire all at once, but that means that up to that point, for the previous 5 minutes we did zero mutations. As opposed to moving each stop context from one `Vec` to the next hundreds of times.